src/psyclone/tests/psyir/transformations/transformations_test.py

```diff
@@ -444,6 +444,99 @@ def create_loops(body):
     omplooptrans.validate(loops[0])
 
 
+def test_omplooptrans_apply_firstprivate(fortran_reader, fortran_writer):
+    ''' Test applying the OMPLoopTrans in cases where a firstprivate
+    clause it is needed to generate functionally equivalent code than
+    the starting serial version. '''
+
+    # Example with a conditional write and a OMPParallelDoDirective
+    psyir = fortran_reader.psyir_from_source(f'''
+        subroutine my_subroutine()
+            integer :: ji, jj, jk, jpkm1, jpjm1, jpim1, scalar1, scalar2
+            real, dimension(10, 10, 10) :: zwt, zwd, zwi, zws
+            do jk = 2, jpkm1, 1
+              do jj = 2, jpjm1, 1
+                do ji = 2, jpim1, 1
+                   if (.true.) then
+                      scalar1 = zwt(ji,jj,jk)
+                   endif
+                   scalar2 = scalar1 + zwt(ji,jj,jk)
+                   zws(ji,jj,jk) = scalar2
+                enddo
+              enddo
+            enddo
+        end subroutine''')
+    omplooptrans = OMPParallelLoopTrans()
+    loop = psyir.walk(Loop)[0]
+    omplooptrans.apply(loop)
+    expected = '''\
+!$omp parallel do default(shared), private(ji,jj,jk,scalar2), \
+firstprivate(scalar1), schedule(auto)
+do jk = 2, jpkm1, 1
+  do jj = 2, jpjm1, 1
+    do ji = 2, jpim1, 1
+      if (.true.) then
+        scalar1 = zwt(ji,jj,jk)
+      end if
+      scalar2 = scalar1 + zwt(ji,jj,jk)
+      zws(ji,jj,jk) = scalar2
+    enddo
+  enddo
+enddo
+!$omp end parallel do\n'''
+
+    gen = fortran_writer(psyir.children[0].children[0])
+    assert expected == gen
+
+
+@pytest.mark.xfail(reason="Issue #598: Firstprivate still misses some cases")
```

Comment:
Rupert and I have been finding that having an 'xfail' at this level can result in the masking of other problems. Please could you move it inside the test and trigger it if the `firstprivate` clause isn't found. i.e.:

    if expected != gen:
        pytest.xfail("...")

Reply:
I moved it inside. Note that the remaining issue is not that the firstprivate is not found, is that the DependencyAnalysis thinks it is a reduction and produces a TransformationError.

In fact I realized that with "force":True the test succeeds, so I moved the string comparison to the previous test and transform it with the force option.